### PR TITLE
Add ExecutorRequirement creation representing creating new

### PR DESCRIPTION
**Pull request by gabrielhartmann**
Executors and using existing Executors

### CODE REVIEW

File: src/main/java/org/apache/mesos/offer/ExecutorRequirement.java [hunk 1]
```diff
@@ -16,13 +16,48 @@
     private ExecutorInfo executorInfo;
     private Collection<ResourceRequirement> resourceRequirements;
 
-    public ExecutorRequirement(ExecutorInfo unverifiedExecutorInfo)
+    /**
+     * This method generates one of two possible Executor requirements.  In the first case, if an empty Executor name is
```
Comment (nickbp) [inline src/main/java/org/apache/mesos/offer/ExecutorRequirement.java]: Empty ID, not name

File: src/main/java/org/apache/mesos/offer/ExecutorRequirement.java [hunk 2]
```diff
@@ -16,13 +16,48 @@
...
+    }
+
+    private ExecutorRequirement(ExecutorInfo executorInfo)
             throws InvalidRequirementException {
-        validateExecutorInfo(unverifiedExecutorInfo);
-        // ExecutorID is always overwritten with a new UUID, even if already present:
-        this.executorInfo = ExecutorInfo.newBuilder(unverifiedExecutorInfo)
-                .setExecutorId(ExecutorUtils.toExecutorId(unverifiedExecutorInfo.getName()))
-                .build();
+        validateExecutorInfo(executorInfo);
```
Comment (nickbp) [inline src/main/java/org/apache/mesos/offer/ExecutorRequirement.java]: Subjective: Validate call could be moved into new create() factory?

**Comment by nickbp** [comment]
Would it be feasible to defer population of new executor IDs in the requirement until after the offer evaluator is run?

The offer evaluator could then use executor ID presence as an explicit signal to pair against a specific executor (and log when they don't match). This is currently being done implicitly via the resource label matching, but without an explicit check the task can still be launched despite the executor ID mismatch. Adding an explicit check with logging could make it easier to discover mistaken changes to the executor ID.

File: src/main/java/org/apache/mesos/offer/ExecutorRequirement.java (hunk 2 above)
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/offer/ExecutorRequirement.java]: I like it in the constructor.  Someday someone may change how these are created.  Putting it in the constructor makes sure whatever path is taken to ultimate construction, validation still takes place.  I considered the approach you're describing.

File: src/main/java/org/apache/mesos/offer/ExecutorRequirement.java (hunk 1 above)
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/offer/ExecutorRequirement.java]: Thanks.

**Comment by gabrielhartmann** [comment]
Your points about the ExecutorRequirements and population of the ExecutorID are good.  I had planned on doing as you describe later, but now is as good a time as any.  I'll do it now.

File: src/main/java/org/apache/mesos/offer/OfferEvaluator.java [hunk 3]
```diff
@@ -310,6 +331,10 @@ private TaskInfo getFulfilledTaskInfo(
               ExecutorInfo.newBuilder(execInfo)
                       .clearResources();
 
+      if (execInfo.getExecutorId().getValue().isEmpty()) {
+        execBuilder.setExecutorId(ExecutorUtils.toExecutorId(execInfo.getName()));
+      }
+
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/offer/OfferEvaluator.java]: This isn't the right place for this.  I'm going to move it up the ExecutorReqirement fulfillment section.

Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/offer/OfferEvaluator.java]: Hmmm... maybe not.

**Pull request merged**